### ISSUE TITLE
`expect` attribute supressing lint, but then saying it's not fulfilled

### ISSUE THREAD

**Issue by asquared31415** [rust-lang/rust-clippy#12998]
**NOTE:** This is using `cargo clippy`, but the actual issue may be in the `lint_reasons` feature.  If this is not appropriate for the rustc repo, let me know and I can copy it over to clippy's.

I tried this code:

```rust
#![feature(lint_reasons)]

fn main() {
    if true {
        #[expect(clippy::needless_return)]
        return;
    }
}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3c57628a2f1df5d48acf2cee23673011) (Use Clippy)

I expected to see this happen: No output because clippy emits the `clippy::needless_return` lint, which should satisfy the `expect`.

Instead, this happened:
```
warning: this lint expectation is unfulfilled
 --> src/main.rs:5:18
  |
5 |         #[expect(clippy::needless_return)]
  |                  ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
  ```

### Meta
`rustc --version --verbose`:
```
rustc 1.77.0-nightly (2df6406b8 2023-12-26)
binary: rustc
commit-hash: 2df6406b886757a3c1475957660a3a4ae6c786de
commit-date: 2023-12-26
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

@rustbot label +T-compiler +A-clippy +A-lint +F-lint_reasons

**Comment by xFrednet** [comment]
Lol, yep found the corporate:

```rs
if !cx.tcx.hir().attrs(expr.hir_id).is_empty() {
    return;
}
```

The lint is apparently suppressed if there is any attribute on the return. Now comes the fun part of figuring out why xD

**Comment by xFrednet** [comment]
Ignoring attributes seem to be intentional: https://github.com/rust-lang/rust-clippy/issues/9361

I can still make it work for `#[expect]` attrs:
